examples/ExactlyOnce/Program.cs

```diff
@@ -15,788 +15,698 @@
+                    break;
+                case "reverse_words":
+                    await new WordReverser().Execute(configuration, transactionalId, adminClient, cts);
+                    break;
+                default:
+                    logger.LogError("Unknown command");
+                    break;
+            }
+
+            completedCancellationTokenSource.Cancel();
+        }
+    }
 
+    public interface ITransactionalProcess<KC, VC, KP, VP>
+    {
         /// <summary>
-        ///     The name of the input Kafka topic containing the lines of text to count the words in.
-        ///         * key = null
-        ///         * value = a line of text (type: string)
+        ///     Reads messages from the consumer,
+        ///     or from something else if no consumer was configured,
+        ///     applies required computation on that data and returns the list
+        ///     of messages to be sent to the output topic.
+        ///     
+        ///     Any exception thrown will cause a rewind of the consumer
+        ///     and an abort of current transaction.
         /// </summary>
-        const string Topic_InputLines = "lines";
+        public List<Message<KP, VP>> ProcessMessages(IConsumer<KC, VC> consumer);
 
         /// <summary>
-        ///     The name of the Kafka topic containing the stream of words to count.
-        ///         * key = individual words (type: string)
-        ///         * value = null.
-        ///
-        ///     Note: This topic is not compacted, so setting the value to null is appropriate. Be
-        ///           aware that for compacted topics though, null has special meaning - it's a
-        ///           tombstone marker.
+        ///     Given current batch of messages to be sent to 
+        ///     the output topic, returns true if it's the moment to commit the
+        ///     transaction.
         /// </summary>
-        const string Topic_Words = "words";
+        public bool IsCommitNeeded(List<Message<KP, VP>> batch);
 
         /// <summary>
-        ///     The name of the Kafka topic containing the word count state (should be compacted).
-        ///         * key = word (type: string)
-        ///         * value = count (type: int)
+        ///     Called at the end of transaction with the Kafka batch 
+        ///     and the <paramref name="committed" /> param that says if
+        ///     the transaction completed successfully. Here the application
+        ///     must commit or abort any other transactions it has
+        ///     ongoing.
         /// </summary>
-        const string Topic_Counts = "counts";
+        public void EndTransaction(List<Message<KP, VP>> batch, bool committed);
+    }
 
+    /// <summary>
+    ///     This transactional processor allows to use a producer an consumer for transactions,
+    ///     by recreating the producer when a fatal error happens, sending offsets to transaction
```

Comment:
i'm unsure if we want to recreate the producer always. it's kind of papering over a real problem, it entered a fatal state for a reason. maybe a config property controls this behavior, and that makes it ok.